src/engine/fileOpsIns.cpp

```diff
@@ -1182,6 +1402,372 @@ void DivEngine::loadOPM(SafeReader& reader, std::vector<DivInstrument*>& ret, St
+    total += (op.sus = ((characteristics >> 5) & 0x1));
+    total += (op.vib = ((characteristics >> 6) & 0x1));
+    total += (op.am = ((characteristics >> 7) & 0x1));
+    total += (op.tl = keyScaleLevel & 0x3F);
+    total += (op.ksl = ((keyScaleLevel >> 6) & 0x3));
+    total += (op.ar = ((attackDecay >> 4) & 0xF));
+    total += (op.dr = attackDecay & 0xF);
+    total += (op.rr = sustainRelease & 0xF);
+    total += (op.sl = ((sustainRelease >> 4) & 0xF));
+    total += (op.ws = waveSelect);
+    return total;
+  };
+
+  auto doParseWoplInstrument = [&](bool isPerc, midibank_t*& metadata, int patchNum) {
+    DivInstrument* ins = new DivInstrument;
+    try {
+      long patchSum = 0;
+      ins->type = DIV_INS_OPL;
+
+      // Establish if it is a blank instrument.
+      String insName = reader.readString(32);
+      patchSum += insName.size();
+      
+      // TODO adapt MIDI key offset to transpose?
+      reader.seek(7, SEEK_CUR);  // skip MIDI params
+      uint8_t instTypeFlags = reader.readC();  // [0EEEDCBA] - see WOPL/OPLI spec
+
+      bool is_4op = ((instTypeFlags & 0x1) == 1);
+      bool is_2x2op = (((instTypeFlags>>1) & 0x1) == 1);
+      bool is_rhythm = (((instTypeFlags>>4) & 0x7) > 0);
+
+      uint8_t feedConnect = reader.readC();
+      uint8_t feedConnect2nd = reader.readC();
+
+      ins->fm.alg = (feedConnect & 0x1);
+      ins->fm.fb = ((feedConnect>>1) & 0xF);
+
+      if (is_4op && !is_2x2op) {
+        ins->fm.ops = 4;
+        ins->fm.alg = (feedConnect & 0x1) | ((feedConnect2nd & 0x1) << 1);
+        for (int i : {2,0,3,1}) { // omfg >_<
+          patchSum += readWoplOp(reader, ins->fm.op[i]);
+        }
+      } else {
+        ins->fm.ops = 2;
+        for (int i : {1,0}) {
+          patchSum += readWoplOp(reader, ins->fm.op[i]);
+        }
+        if (is_rhythm) {
+          ins->fm.opllPreset = (uint8_t)(1<<4);
+        } else if (is_2x2op) {
+          // Note: Pair detuning offset not mappable. Use E5xx effect :P
+          ins->name = stringNotBlank(insName)
+            ? fmt::sprintf("%s (1)", insName)
+            : fmt::sprintf("%s[%s] %s Patch %d (1)",
+              stripPath, metadata->name, (isPerc) ? "Drum" : "Melodic", patchNum);
+          insList.push_back(ins);
+          patchSum = 0;
+          ins = new DivInstrument;
+          ins->type = DIV_INS_OPL;
+          ins->name = fmt::sprintf("%s (2)", insName);
+          for (int i : {1,0}) {
+            patchSum += readWoplOp(reader, ins->fm.op[i]);
+          }
+        }
+
+        if (!is_2x2op) {
+          reader.seek(10, SEEK_CUR); // skip unused operator pair
+        }
+      }
+
+      if (version >= 3) {
+        reader.readS_BE(); // skip keyon delay
+        reader.readS_BE(); // skip keyoff delay
+      }
+
+      if (patchSum > 0) {
+        // Write instrument
+        // TODO: OPL3BankEditor hardcodes GM1 Melodic patch names which are not included in the bank file......
```

Comment:
P.S. You may look at this and related files around: <https://github.com/Wohlstand/OPL3BankEditor/blob/master/src/ins_names_data.h>

Reply:
Yeah I did see this. I think for now having bank+patch as fallback is enough for now here. If adding hardcoded names end up being a requewted feature I'll shove it in at a later stage. :)

Reply:
Hardcoded names absolutely do not require, just a hint at the bank editor itself. They were made for bank formats that lack the ability to store the instrument name at all.

Reply:
Yeah. I think for our purposes given Furnace is (at present) not a General MIDI tool, the need for these aren't too important in my view. This may change however depending on user demand - the joys of FOSS. :)